### PR TITLE
Fix crash with type parameter on type constrained to file-local type

### LINKED ISSUE #62435: Specifying `file` class as generic constraints crashes the compiler with NRE
```csharp
using System;
file class C
{

}


public class D<T> where T : C { } // crash happens also if `public` is replaced with `file`.
```

Seen on SharpLab:

```csharp
System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.VisitType[T](TypeWithAnnotations typeWithAnnotationsOpt, TypeSymbol type, Func`4 typeWithAnnotationsPredicate, Func`4 typePredicate, T arg, Boolean canDigThroughNullable, Boolean useDefaultType, Boolean visitCustomModifiers) in /_/src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:line 692
   at Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.IsFileTypeOrUsesFileTypes(TypeSymbol type) in /_/src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:line 1366
   at Microsoft.CodeAnalysis.CSharp.Binder.CheckConstraintTypeVisibility(Symbol containingSymbol, Location location, TypeWithAnnotations constraintType, BindingDiagnosticBag diagnostics) in /_/src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:line 420
   at Microsoft.CodeAnalysis.CSharp.Binder.RemoveInvalidConstraints(TypeParameterSymbol typeParameter, TypeParameterConstraintClause constraintClause, ArrayBuilder`1 syntaxNodesOpt, Boolean performOnlyCycleSafeValidation, BindingDiagnosticBag diagnostics) in /_/src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:line 381
   at Microsoft.CodeAnalysis.CSharp.Binder.RemoveInvalidConstraints(ImmutableArray`1 typeParameters, ArrayBuilder`1 constraintClauses, ArrayBuilder`1 syntaxNodes, Boolean performOnlyCycleSafeValidation, BindingDiagnosticBag diagnostics) in /_/src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:line 349
   at Microsoft.CodeAnalysis.CSharp.Binder.BindTypeParameterConstraintClauses(Symbol containingSymbol, ImmutableArray`1 typeParameters, TypeParameterListSyntax typeParameterList, SyntaxList`1 clauses, BindingDiagnosticBag diagnostics, Boolean performOnlyCycleSafeValidation, Boolean isForOverride) in /_/src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:line 102
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.MakeTypeParameterConstraintTypes(BindingDiagnosticBag diagnostics) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:line 311
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetTypeParameterConstraintTypes() in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:line 281
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetTypeParameterConstraintTypes(Int32 ordinal) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:line 265
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceTypeParameterSymbol.ResolveBounds(ConsList`1 inProgress, BindingDiagnosticBag diagnostics) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:line 582
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceTypeParameterSymbolBase.GetBounds(ConsList`1 inProgress) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:line 239
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceTypeParameterSymbolBase.GetConstraintTypes(ConsList`1 inProgress) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:line 94
   at Microsoft.CodeAnalysis.CSharp.Symbols.TypeParameterSymbol.EnsureAllConstraintsAreResolved(ImmutableArray`1 typeParameters) in /_/src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs:line 376
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceTypeParameterSymbolBase.EnsureAllConstraintsAreResolved() in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:line 213
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceTypeParameterSymbolBase.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs:line 358
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:line 567
   at Microsoft.CodeAnalysis.CSharp.Symbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Symbol.cs:line 827
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.<>c__DisplayClass63_1.<ForceComplete>b__0(Int32 i) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol_Completion.cs:line 59
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass6_0`1.<WithCurrentUICulture>b__0(T param) in /_/src/Compilers/Core/Portable/InternalUtilities/UICultureUtilities.cs:line 171
   at Roslyn.Utilities.RoslynParallel.<>c__DisplayClass1_0.<For>g__errorHandlingBody|0(Int32 i) in /_/src/Compilers/Core/Portable/InternalUtilities/RoslynParallel.cs:line 43
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body)
   at Roslyn.Utilities.RoslynParallel.For(Int32 fromInclusive, Int32 toExclusive, Action`1 body, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/InternalUtilities/RoslynParallel.cs:line 23
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol_Completion.cs:line 56
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceModuleSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs:line 225
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs:line 902
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:line 2946
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnosticsWithoutFiltering(CompilationStage stage, Boolean includeEarlierStages, BindingDiagnosticBag builder, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:line 2721
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnostics(CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:line 2633
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetAllDiagnosticsWithoutStateTrackingAsync(ImmutableArray`1 analyzers, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:line 478
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetAllDiagnosticsAsync(CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:line 326
   at MirrorSharp.Internal.Roslyn.RoslynSession.GetDiagnosticsAsync(CancellationToken cancellationToken) in D:\a\SharpLab\SharpLab\!roslyn-branches\dotnet-main\sharplab\source\#external\mirrorsharp\Common\Internal\Roslyn\RoslynSession.cs:line 102
   at MirrorSharp.Internal.Handlers.SlowUpdateHandler.ExecuteAsync(AsyncData data, WorkSession session, ICommandResultSender sender, CancellationToken cancellationToken) in D:\a\SharpLab\SharpLab\!roslyn-branches\dotnet-main\sharplab\source\#external\mirrorsharp\Common\Internal\Handlers\SlowUpdateHandler.cs:line 39
   at MirrorSharp.Internal.Connection.ReceiveAndProcessInternalAsync(CancellationToken cancellationToken) in D:\a\SharpLab\SharpLab\!roslyn-branches\dotnet-main\sharplab\source\#external\mirrorsharp\Common\Internal\Connection.cs:line 118
   at MirrorSharp.Internal.Connection.ReceiveAndProcessAsync(CancellationToken cancellationToken) in D:\a\SharpLab\SharpLab\!roslyn-branches\dotnet-main\sharplab\source\#external\mirrorsharp\Common\Internal\Connection.cs:line 73

```

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #60819 
Fixes #62435

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs [hunk 1]
```diff
@@ -411,7 +411,7 @@ private TypeParameterConstraintClause GetDefaultTypeParameterConstraintClause(Ty
                 diagnostics.Add(ErrorCode.ERR_BadVisBound, location, containingSymbol, constraintType.Type);
             }
 
-            if (constraintType.Type.IsFileTypeOrUsesFileTypes() && !containingSymbol.ContainingType.IsFileTypeOrUsesFileTypes())
+            if (constraintType.Type.IsFileTypeOrUsesFileTypes() && !(containingSymbol as TypeSymbol ?? containingSymbol.ContainingType).IsFileTypeOrUsesFileTypes())
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:414]: doc?

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs [hunk 2]
```diff
@@ -2074,6 +2075,28 @@ class E
             Diagnostic(ErrorCode.ERR_FileTypeDisallowedInSignature, "C").WithArguments("C", "E.M<T>(T)").WithLocation(10, 30));
     }
 
+    [Fact, WorkItem(62435, "https://github.com/dotnet/roslyn/issues/62435")]
+    public void Constraints_02()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:2079]: Consider also covering type constraint on local function.
Consider covering all occurrences of type constraints (classes, structs, interfaces, records, delegate types) by parameterizing this test.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs [hunk 3]
```diff
@@ -411,7 +411,9 @@ private TypeParameterConstraintClause GetDefaultTypeParameterConstraintClause(Ty
                 diagnostics.Add(ErrorCode.ERR_BadVisBound, location, containingSymbol, constraintType.Type);
             }
 
-            if (constraintType.Type.IsFileTypeOrUsesFileTypes() && !containingSymbol.ContainingType.IsFileTypeOrUsesFileTypes())
+            // Whether a member is effectively file-local is determined on the type level. When non-type members have constraints we need to check the containing symbol.
+            // If the 'containingSymbol.ContainingSymbol' is not a type, then 'containingSymbol' is not a member, and we don't need to enforce file types here.
+            if (constraintType.Type.IsFileTypeOrUsesFileTypes() && (containingSymbol as TypeSymbol ?? containingSymbol.ContainingSymbol as TypeSymbol)?.IsFileTypeOrUsesFileTypes() == false)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:416]: >(containingSymbol as TypeSymbol ?? containingSymbol.ContainingSymbol as TypeSymbol) [](http://example.com/codeflow?start=67&length=83)

This is getting harder to read.
I'd extract this to a local function or local variable and maybe do it more formally as a switch case: for types, we check the type; for methods, we check the containing type; for any other symbol (local function), there's nothing to check.
What do you think?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3), but I wonder if we can make the code more readable

File: src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs [hunk 4]
```diff
@@ -411,9 +411,21 @@ private TypeParameterConstraintClause GetDefaultTypeParameterConstraintClause(Ty
...
+                // if the containing symbol of the constraint is a member, we need to ensure the nearest containing type is a file type.
+                var possibleFileType = containingSymbol switch
+                {
+                    TypeSymbol typeSymbol => typeSymbol,
+                    LocalFunctionSymbol => null,
+                    MethodSymbol method => (TypeSymbol)method.ContainingSymbol,
+                    _ => throw ExceptionUtilities.UnexpectedValue(containingSymbol)
+                };
+                Debug.Assert(possibleFileType?.IsDefinition != false);
+                if (possibleFileType?.IsFileTypeOrUsesFileTypes() == false)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:425]: >if (possibleFileType?.IsFileTypeOrUsesFileTypes() == false) [](http://example.com/codeflow?start=16&length=59)

Is this check meant to replace the `&& (containingSymbol as TypeSymbol ?? ... )?.IsFileTypeOrUsesFileTypes() == false` above?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs:425]: Thanks, looks like I forgot to delete the original check.

**Review by cston — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5)

**Pull request merged**